cocos/renderer/frame-graph/FrameGraph.h

```diff
@@ -0,0 +1,145 @@
+/****************************************************************************
+ Copyright (c) 2021 Xiamen Yaji Software Co., Ltd.
+
+ http://www.cocos.com
+
+ Permission is hereby granted, free of charge, to any person obtaining a copy
+ of this software and associated engine source code (the "Software"), a limited,
+ worldwide, royalty-free, non-assignable, revocable and non-exclusive license
+ to use Cocos Creator solely to develop games on your target platforms. You shall
+ not use Cocos Creator software for developing other software or tools that's
+ used for developing games. You are not granted to publish, distribute,
+ sublicense, and/or sell copies of Cocos Creator.
+
+ The software or tools in this License Agreement are licensed, not sold.
+ Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+****************************************************************************/
+
+#pragma once
+
+#include "Blackboard.h"
+#include "DevicePass.h"
+#include "CallbackPass.h"
+#include "PassNode.h"
+#include "PassNodeBuilder.h"
+#include "ResourceEntry.h"
+#include "ResourceNode.h"
+#include <cstdint>
+#include <string>
+#include <vector>
+
+namespace cc {
+namespace framegraph {
+
+class FrameGraph final {
+public:
+    using ResourceHandleBlackboard = Blackboard<StringHandle, Handle::IndexType, Handle::UNINITIALIZED>;
+
+    FrameGraph()                   = default;
+    ~FrameGraph();
+    FrameGraph(FrameGraph const &) = delete;
+    FrameGraph(FrameGraph &&)      = delete;
+    FrameGraph &operator=(FrameGraph const &) = delete;
+    FrameGraph &operator=(FrameGraph &&) = delete;
+
+    static StringHandle stringToHandle(char const *const name) noexcept;
+    static char const * handleToString(StringHandle const handle) noexcept;
```

Comment:
Why use char* instead of String?

Reply:
for the names are almost always literals, and the string pool maintains them as chunks of raw memory.

Reply:
But if it is passed by `string.c_str()`, then it is strange.

Reply:
Sorry, i mean it is dangerous as the string may be destroyed.

Reply:
the string pool makes an internal copy of every enlisted string, so it is safe here